Added implementation of the video interface to dailymotion

wassgha opened the pull request:
This implements video interface inside the amp-dailymotion extension

### Changes

- Reverse-engineered the Dailymotion API (added action/status messages list)
- Implemented video interface methods
- Added autoplay example to the amp-dailymotion example
- Added manual controls example to the amp-dailymotion example
- Added amp-dailymotion to the players integration tests
- Minor corrections in amp-youtube
- Added documentation

Closes #6285

aghassemi commented on test/integration/test-video-players-helper.js:
```diff
@@ -65,22 +65,24 @@ export function runVideoPlayerIntegrationTests(
         const pauseButton = createButton(r, 'pause');
         const muteButton = createButton(r, 'mute');
         const unmuteButton = createButton(r, 'unmute');
+        // Dailymotion doesn't support mute/unmute when video is paused,
```
If that's the case, we can't support daily motion as autoplay on mobile only works if video is muted, so before we call `play`, `mute` must be honored. Might be worth investigating to see if there is any workaround for this.

wassgha commented:
Amended commit
### Changes
- Added documentation
- Added manual event firing for mute/unmute when video didn't start yet
- Reverted test order to mute first

aghassemi commented on extensions/amp-dailymotion/0.1/amp-dailymotion.js:
```diff
@@ -16,16 +16,59 @@
     /** @private {?Element} */
     this.iframe_ = null;
+
+    /** @private {?boolean}  */
```
just `{boolean}` I assume it is non-nullable

aghassemi commented on extensions/amp-dailymotion/0.1/amp-dailymotion.js:
```diff
@@ -34,8 +77,21 @@ class AmpDailymotion extends AMP.BaseElement {
+
+  /** @override */
+  isInteractive() {
+    // YouTube videos are always interactive. There is no YouTube param that
```
update comments, also let's make sure there actually isn't a dailymotion param that can completely hide the UI controls.

aghassemi commented on extensions/amp-dailymotion/0.1/amp-dailymotion.js:
```diff
@@ -44,21 +100,54 @@ class AmpDailymotion extends AMP.BaseElement {
   }
 
   /** @override */
+  renderOutsideViewport() {
```
it's fine to keep this as dailymotion embed seems pretty heavy but update the comments.

aghassemi commented on extensions/amp-dailymotion/0.1/amp-dailymotion.js:
```diff
@@ -44,21 +100,54 @@ class AmpDailymotion extends AMP.BaseElement {
+    );
+
+    const loaded = this.loadPromise(this.iframe_).then(() => {
+      this.element.dispatchCustomEvent(VideoEvents.LOAD);
```
let's do
```
return this.playerReadyPromise_.then( () => {
   this.element.dispatchCustomEvent(VideoEvents.LOAD);
});
```

aghassemi commented on extensions/amp-dailymotion/0.1/amp-dailymotion.js:
```diff
@@ -70,6 +159,66 @@ class AmpDailymotion extends AMP.BaseElement {
+      this.playerReadyResolver_(true);
+    }
+
+    console.log(data);
```
remove

aghassemi commented on extensions/amp-dailymotion/0.1/amp-dailymotion.js:
```diff
@@ -70,6 +159,66 @@ class AmpDailymotion extends AMP.BaseElement {
+          (data.volume == 0 || data.muted == 'true') ?
+          VideoEvents.MUTED : VideoEvents.UNMUTED
+        );
+        this.element.dispatchCustomEvent(muteEvent);
```
we need to make sure we only dispatch this when muted state changed. I think this will dispatch multiple `UNMUTED` events when user changes volume for example. amp-youtube keeps current muted state and compares before dispatching, we can do the same here.

aghassemi commented on extensions/amp-dailymotion/0.1/amp-dailymotion.js:
```diff
@@ -96,10 +245,63 @@ class AmpDailymotion extends AMP.BaseElement {
   pauseCallback() {
-    if (this.iframe_ && this.iframe_.contentWindow) {
-      this.iframe_.contentWindow./*OK*/postMessage('pause', '*');
+    if (this.iframe_ && this.iframe_.contentWindow
```
I don't think this needs the guards. `this.pause()` should just work (YT had a bug with calling pause when video was not playing)

wassgha replied:
For some reason git doesn't think this was fixed, but it was fixed in the last commit.

aghassemi replied:
I can still see the `if`

wassgha commented:
Couldn't find a better solution for the "play() can only be initiated by a user gesture" problem, I had to call play twice if the video has the autoplay attribute.

aghassemi commented:
@wassgha Great. That's a pretty good workaround, not too hacky and should be pretty safe.

wassgha commented:
###Changes
- Fixed multiple bugs in dispatching events for mute/unmute
- Fixed autoplay on mobile
- Added JSDoc for various variables
- Whitelisted Dailymotion's dependency on the video manager

honeybadgerdontcare commented on extensions/amp-dailymotion/0.1/validator-amp-dailymotion.protoascii:
```diff
@@ -64,6 +64,7 @@ tags: {  # <amp-dailymotion>
     mandatory: true
     value_regex_casei: "[a-z0-9]+"
   }
+  attrs: { name: "autoplay" }
```
We attempt to keep the attrs in alphabetical order, so please move this up above `data-endscreen-enable`

wassgha replied:
Alright, thanks!

aghassemi commented on extensions/amp-dailymotion/0.1/amp-dailymotion.js:
```diff
@@ -15,17 +15,82 @@
+ * @private
+ */
+
+// Events reverse-engineered from the Dailymotion API
```
merge with js comment above

aghassemi commented on extensions/amp-dailymotion/0.1/amp-dailymotion.js:
```diff
@@ -38,27 +103,71 @@ class AmpDailymotion extends AMP.BaseElement {
+
+  /** @override */
+  isInteractive() {
+    // Dailymotion videos are always interactive. There is no YouTube param that
```
s/YouTube/Dailymotion

aghassemi commented on extensions/amp-dailymotion/0.1/amp-dailymotion.js:
```diff
@@ -38,27 +103,71 @@ class AmpDailymotion extends AMP.BaseElement {
+  /** @override */
+  isInteractive() {
+    // Dailymotion videos are always interactive. There is no YouTube param that
+    // makes the video non-interactive. Even data-param-control=0 will not
```
I don't think data-param-controls exists for dailymotion, remove sentence.

wassgha replied:
There's controls = false but it still allows play/pause by clicking the video

aghassemi commented on extensions/amp-dailymotion/0.1/amp-dailymotion.js:
```diff
@@ -96,10 +271,72 @@ class AmpDailymotion extends AMP.BaseElement {
+    // Hack to solve autoplay problem on Chrome Android
+    // (first play always fails)
+    if (isAutoplay && this.playerState_ != DailymotionEvents.PAUSE) {
+      this.loadedPromise_.then(() => {
```
can this be `playerReadyPromise_` so we can remove the whole `loadedPromise_`?

wassgha replied:
Actually this is part of the hack, `loadedPromise` resolves when the video starts buffering (not when the iframe finishes loading), that way we send a first play event to start buffering then a second play event to actually start playing

wassgha replied:
I renamed the event from `DailymotionEvents. LOADED` to `DailymotionEvents. STARTED_BUFFERING` in the last commit

Gregable commented:
These validator changes should be live everywhere within an hour.
